posthog/models/activity_logging/model_activity.py

```diff
@@ -0,0 +1,54 @@
+from threading import local
+from django.core.handlers.wsgi import WSGIRequest
+from typing import Any
+from django.db import models
+from posthog.models.signals import model_activity_signal
+from loginas.utils import is_impersonated_session
+
+_thread_local = local()
+
+
+def get_current_request():
+    return getattr(_thread_local, "request", None)
+
+
+def set_current_request(request):
+    _thread_local.request = request
+
+
+class ModelActivityMixin(models.Model):
+    """
+    A mixin that automatically sends activity signals when a model is created or updated.
+    The model's class name will be used as the scope for activity logging.
+    """
+
+    def save(self, *args: Any, **kwargs: Any) -> None:
+        # Get a copy of the existing instance before saving
+        if self.pk:
+            before_update = self.__class__.objects.filter(pk=self.pk).first()
+            if before_update:
+                before_update._state.adding = False  # Ensure the copy knows it's not a new instance
+                before_update.pk = before_update.pk  # Ensure pk is copied
+        else:
+            before_update = None
+
+        change_type = "updated" if self.pk else "created"
+
+        # Get request from thread local
+        request = get_current_request()
```

Comment:
i guess folk could forget to set this (or not know it's required) and we should give some helpful error if they do

Reply:
Yeah, good call. This is the most brittle part of this approach. But I noticed that feature flag scheduled changes set the request user correctly, so we must be doing this sort of thing elsewhere too.

Reply:
Actually, it's only required if you were impersonating. I think I should make it more explicit by making you set `was_impersonated` in tls rather than passing the `request`.